File: DataBinding/PageAndRDL/JsonDataSource/VB.NET/Web Service/BasicAuthHttpModule.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Security.Principal
Imports System.Text
Imports System.Threading
Imports System.Web



Public Class BasicAuthHttpModule : Implements IHttpModule

	Private Const Realm As String = "My Realm"

	Public Sub Init(ByVal context As HttpApplication) Implements IHttpModule.Init
		'Register event handlers
		AddHandler context.AuthenticateRequest, AddressOf OnApplicationAuthenticateRequest
		AddHandler context.EndRequest, AddressOf OnApplicationEndRequest
	End Sub

	Private Shared Sub SetPrincipal(principal As IPrincipal)
		Thread.CurrentPrincipal = principal
		If Not IsNothing(HttpContext.Current) Then
			HttpContext.Current.User = principal
		End If
	End Sub

	' TODO: Here is where you would validate the username and password.
	Private Shared Function CheckPassword(username As String, password As String) As Boolean
		Return username = "admin" And password = "1"
	End Function

	Private Shared Sub AuthenticateUser(credentials As String)
		Try
			Dim encoding As Encoding = encoding.GetEncoding("iso-8859-1")
			credentials = encoding.GetString(Convert.FromBase64String(credentials))

			Dim separator As Integer = credentials.IndexOf(":")
			Dim name As String = credentials.Substring(0, separator)
			Dim password As String = credentials.Substring(separator + 1)

			If (CheckPassword(name, password)) Then
				Dim identity = New GenericIdentity(name)
				SetPrincipal(New GenericPrincipal(identity, Nothing))
			Else
				' Invalid username or password.
				HttpContext.Current.Response.StatusCode = 403
			End If
		Catch ex As Exception
			' Credentials were not formatted correctly.
			HttpContext.Current.Response.StatusCode = 401
		End Try
	End Sub
	' http://cacheandquery.com/blog/2011/03/customizing-asp-net-mvc-basic-authentication/
	Private Sub OnApplicationAuthenticateRequest(sender As Object, e As EventArgs)
		Dim request = HttpContext.Current.Request
		Dim info As System.IO.FileInfo = New System.IO.FileInfo(request.Url.AbsolutePath)
		If not info.Name.Equals("GetJson") Then Return

		Dim authHeader As String = request.Headers("Authorization")
		' RFC 2617 sec 1.2, "scheme" name is case-insensitive
		If ((Not (authHeader) Is Nothing) _
			AndAlso (authHeader.StartsWith("basic ", StringComparison.OrdinalIgnoreCase) _
			AndAlso (authHeader.Length > 6))) Then
			AuthenticateUser(authHeader.Substring(6))
		Else
			HttpContext.Current.Response.StatusCode = 401
		End If
	End Sub

	' If the request was unauthorized, add the WWW-Authenticate header 
	' to the response.
	Private Sub OnApplicationEndRequest(sender As Object, e As EventArgs)
		Dim response = HttpContext.Current.Response
		If (response.StatusCode = 401) Then
			response.Headers.Add("WWW-Authenticate", String.Format("Basic realm=""{0}""", Realm))
		End If
	End Sub

	Public Sub Dispose() Implements IHttpModule.Dispose
	End Sub

End Class

File: DataBinding/PageAndRDL/JsonDataSource/C#/Web Service/BasicAuthHttpModule.cs
using System;
using System.Security.Principal;
using System.Text;
using System.Threading;
using System.Web;

namespace WebService
{
	public class BasicAuthHttpModule : IHttpModule
	{
		private const string Realm = "My Realm";

		public void Init(HttpApplication context)
		{
			// Register event handlers
			context.AuthenticateRequest += OnApplicationAuthenticateRequest;
			context.EndRequest += OnApplicationEndRequest;
		}

		private static void SetPrincipal(IPrincipal principal)
		{
			Thread.CurrentPrincipal = principal;
			if (HttpContext.Current != null)
			{
				HttpContext.Current.User = principal;
			}
		}

		// TODO: Here is where you would validate the username and password.
		private static bool CheckPassword(string username, string password)
		{
			return username == "admin" && password == "1";
		}

		private static void AuthenticateUser(string credentials)
		{
			try
			{
				var encoding = Encoding.GetEncoding("iso-8859-1");
				credentials = encoding.GetString(Convert.FromBase64String(credentials));

				int separator = credentials.IndexOf(':');
				string name = credentials.Substring(0, separator);
				string password = credentials.Substring(separator + 1);

				if (CheckPassword(name, password))
				{
					var identity = new GenericIdentity(name);
					SetPrincipal(new GenericPrincipal(identity, null));
				}
				else
				{
					// Invalid username or password.
					HttpContext.Current.Response.StatusCode = 403;
				}
			}
			catch (FormatException)
			{
				// Credentials were not formatted correctly.
				HttpContext.Current.Response.StatusCode = 401;
			}
		}

		// http://cacheandquery.com/blog/2011/03/customizing-asp-net-mvc-basic-authentication/
		private static void OnApplicationAuthenticateRequest(object sender, EventArgs e)
		{
			var request = HttpContext.Current.Request;
			System.IO.FileInfo info = new System.IO.FileInfo(request.Url.AbsolutePath);
			if (!info.Name.Equals("GetJson")) return;

			var authHeader = request.Headers["Authorization"];
			// RFC 2617 sec 1.2, "scheme" name is case-insensitive
			if (authHeader != null && authHeader.StartsWith("basic ", StringComparison.OrdinalIgnoreCase) && authHeader.Length > 6)
			{ 
				AuthenticateUser(authHeader.Substring(6));
			}
			else
			{
				HttpContext.Current.Response.StatusCode = 401;
			}
		}

		// If the request was unauthorized, add the WWW-Authenticate header 
		// to the response.
		private static void OnApplicationEndRequest(object sender, EventArgs e)
		{
			var response = HttpContext.Current.Response;
			if (response.StatusCode == 401)
			{
				response.Headers.Add("WWW-Authenticate",
					string.Format("Basic realm=\"{0}\"", Realm));
			}
		}

		public void Dispose()
		{
		}
	}
}
